prow/gangway/google/client.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package google
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"golang.org/x/oauth2"
+	googleOAuth "golang.org/x/oauth2/google"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+
+	pb "k8s.io/test-infra/prow/gangway"
+)
+
+// This is the client library for Go clients that need to access to the Prow
+// API, aka Gangway, when Gangway is deployed in a GKE cluster and integrated
+// with Cloud Endpoints.
+//
+// Go clients need to always append 2 things to the metadata of a gRPC call:
+//
+//  	1. The JWT (authentication) token (to make the call identify itself as
+//  	an allowlisted client in our api_config_auth.yaml configuration for Cloud
+//  	Endpoints [1]), and
+//
+//  	2. The API key (that is generated by the client's GCP Project).
+//
+//  The JWT token is generated from a GCP Service Account key file (JSON). The
+//  API key is generated from the GCP user interface, like this:
+//  https://cloud.google.com/docs/authentication/api-keys#create.
+//
+//  For us, the clients must supply the service account JSON key file, API key,
+//  audience, and finally the address where Gangway is being served. An example
+//  client application using this library is provided in the Prow codebase under
+//  prow/gangway/example/main.go.
+//
+// [1]: https://github.com/GoogleCloudPlatform/golang-samples/blob/e888c56cb843f475db4f79b391be999518e63db4/endpoints/getting-started-grpc/README.md#configuring-authentication-and-authenticating-requests
+
+type Client struct {
+	// JWT token-based authentication and GCP Project identification.
+	keyBytes    []byte
+	audience    string
+	tokenSource oauth2.TokenSource
+
+	// apiKey identifies the GCP Project.
+	apiKey string
+
+	addr string
+	conn *grpc.ClientConn
+
+	// GRPC is the auto-generated gRPC client interface for gangway.
+	GRPC pb.ProwClient
```

Comment:
Maybe just extend GRPC interface unless you plan to add other protocols.

Reply:
Do you mean struct embedding https://gobyexample.com/struct-embedding ?

I think @ziyulin0 was asking for an HTTP-based client as well, so, maybe we should support that protocol here also. Could be done in a follow-up PR.

Reply:
yes that s what I meant. ok let's check with Ziyu.

Reply:
Let's refactor as necessary in follow-up PRs as this PR is pretty big already.

FWIW this code has been tested to work as expected (can trigger a job and wait for its status).